adagucserverEC/CConvertLatLonBndsData.cpp

```diff
@@ -0,0 +1,285 @@
+ * limitations under the License.
+ *
+ ******************************************************************************/
+
+#include "CConvertLatLonBnds.h"
+#include "CFillTriangle.h"
+#include "CImageWarper.h"
+
+// #define CConvertLatLonBnds_DEBUG
+/**
+ * This function draws the virtual 2D variable into a new 2D field
+ */
+int CConvertLatLonBnds::convertLatLonBndsData(CDataSource *dataSource, int mode) {
+  CDFObject *cdfObject = dataSource->getDataObject(0)->cdfObject;
+  if (!isThisLatLonBndsData(cdfObject)) return 1;
+  size_t nrDataObjects = dataSource->getNumDataObjects();
+  if (nrDataObjects <= 0) return 1;
+
+  CDataSource::DataObject *dataObjects[nrDataObjects];
+  for (size_t d = 0; d < nrDataObjects; d++) {
+    dataObjects[d] = dataSource->getDataObject(d);
+  }
+#ifdef CConvertLatLonBnds_DEBUG
+  CDBDebug("convertLatLonBndsData %s", dataObjects[0]->cdfVariable->name.c_str());
+#endif
+
+  CDF::Variable *destRegularGrid[nrDataObjects];
+  CDF::Variable *irregularGridVar[nrDataObjects];
+
+  // Make references destRegularGrid and irregularGridVar
+  for (size_t d = 0; d < nrDataObjects; d++) {
+    destRegularGrid[d] = dataObjects[d]->cdfVariable;
+    CT::string orgName = destRegularGrid[d]->name.c_str();
+    orgName.concat("_backup");
+    irregularGridVar[d] = cdfObject->getVariableNE(orgName.c_str());
+    if (irregularGridVar[d] == NULL) {
+      CDBError("Unable to find orignal variable with name %s", orgName.c_str());
+      return 1;
+    }
+  }
+  CDF::Variable *longitudeBnds = cdfObject->getVariableNE("lon_bnds");
+  CDF::Variable *latitudeBnds = cdfObject->getVariableNE("lat_bnds");
+
+  for (size_t d = 0; d < nrDataObjects; d++) {
+    dataSource->readVariableDataForCDFDims(irregularGridVar[d], CDF_FLOAT);
+    float fltFill = NC_FILL_FLOAT;
+    double dblFill = NC_FILL_DOUBLE;
+    CDF::Attribute *fillValue = irregularGridVar[d]->getAttributeNE("_FillValue");
+    if (fillValue != NULL) {
+      dataObjects[d]->hasNodataValue = true;
+      fillValue->getData(&dblFill, 1);
+      dataObjects[d]->dfNodataValue = dblFill;
+      fltFill = (float)dblFill;
+      destRegularGrid[d]->getAttribute("_FillValue")->setData(CDF_DOUBLE, &fltFill, 1);
+    } else {
+      dataObjects[d]->hasNodataValue = false;
+    }
+  }
+  float fltFill = NC_FILL_FLOAT;
+  if (dataObjects[0]->hasNodataValue) {
+    fltFill = (float)dataObjects[0]->dfNodataValue;
+  }
+
+  // If the data was not populated in the code above, try to read it from the file
+  if (longitudeBnds->data == nullptr) {
+    longitudeBnds->readData(CDF_DOUBLE, true);
+  }
+  if (latitudeBnds->data == nullptr) {
+    latitudeBnds->readData(CDF_DOUBLE, true);
+  }
+
+  // Detect minimum and maximum values
+  MinMax minMax;
+  minMax = getMinMax(((float *)irregularGridVar[0]->data), dataObjects[0]->hasNodataValue, (double)fltFill, irregularGridVar[0]->getSize());
+  CDBDebug("minMax %f %f", minMax.min, minMax.max);
```

Comment:
Please ensure that there isn't too much logging going on.

Reply:
Done